## Formatierte Ausgabe


### Wege die Ausgabe zu formatieren

<img class="imgright" src="../images/Printer_in_1568.webp" srcset="../images/Printer_in_1568_300w.webp 300w" alt="Drucken im 16. Jahrhundert"/>

In diesem Kapitel unseres Python-Tutorials werden wir uns intensiv mit den verschiedenen Methoden beschäftigen, mit denen man die Ausgaben formatieren bzw. formatierte Strings erzeugen kann. Wir stellen die verschiedenen Arten vor, aber wir empfehlen "formatierte" Stringliterale zu benutzen, die wir am Anfang des Kapitels erklären. Diese Methode ist die bei weitem lesbarste und eleganteste.

Bisher hatten wir die print-Funktion auf zwei Arten benutzt, wenn wir beispielsweise zwei Werte ausgeben wollten:

Die einfachste Art, aber sicherlich nicht die eleganteste:
Wir benutzen print mit einer kommaseparierten Liste von Werten, um die Ergebnisse auszugeben, wie wir im folgenden Beispiel sehen können. Alle Werte werden durch Leerzeichen getrennt, was dem Default-Verhalten entspricht. Wir können die Default-Trennung durch einen beliebigen String ersetzen. Dazu müssen wir diesen String lediglich durch den Schlüsselwort-Parameter "sep" der Print-Funktion ersetzen:

In [1]:
q = 459
p = 0.098
print(q, p, p * q)

459 0.098 44.982


In [55]:
print(q, p, p * q, sep=",")

459,0.098,44.982


In [56]:
print(q, p, p * q, sep=" :-) ")

459 :-) 0.098 :-) 44.982


Alternativ können wir auch aus den Werten mittels der String-Konkatenation einen neuen String konstruieren:

In [57]:
print(str(q) + " " + str(p) + " " + str(p * q))

459 0.098 44.982


Die zweite Methode ist schlechter, - da komplizierter, - als die erste. 

### Formatierte Stringliterale

Formatierte Stringliterale oder f-strings wurden ab Python 3.6 eingeführt. Ihre Syntax ist sehr einfach und flexibel. Ein Beispiel ist:

In [58]:
name,alter ="Jonas",19
f"Mein Freund {name} ist {alter} Jahre alt"


'Mein Freund Jonas ist 19 Jahre alt'

Die Anführungsstriche in der Ausgabe verschwinden, wenn wir den f-string mit print drucken.


In [59]:
name,alter ="Jonas",19
f"Mein Freund {name} ist {alter} Jahre alt"
print(f"Mein Freund {name} ist {alter} Jahre alt")

Mein Freund Jonas ist 19 Jahre alt


In den geschweiften Klammern können nicht nur Variablennamen stehen, sondern auch ganze Ausdrücke, ein Beispiel folgt.

In [60]:
name,alter="Jonas",28
name2,alter2="Karl",32
f"{name} ist {abs(alter-alter2)} Jahre {'jünger' if alter-alter2<0 else 'älter'} als {name2}."

'Jonas ist 4 Jahre jünger als Karl.'

Wir sehen hier, dass auch das ternäre if in f-strings gute Dienste tun kan. Übrigens kann man auch ein grosses "F" verwenden. Achten muss man darauf, dass strings in den geschweiften Klammern mit einfachem Anführungszeichen begrenzt werden, wenn der f-string selber doppelte Anführungszeichen hat oder umgekehrt. Die Verwendung von \ in f-strings ist aber auch möglich(s.u.). Wollen wir ein dictionary ausgeben, müssen wir die geschweiften Klammern des dictionary verdoppeln, um es von den geschweiften Klammern des f-strings abzugrenzen und den Doppelpunkt als Teil des dictionary zu markieren(dazu gleich mehr). 

In [61]:
f"Ein Python dictionary d={{\"a\":1,'b':2}}"

'Ein Python dictionary d={"a":1,\'b\':2}'

Auch ein Funktionsaufruf kann natürlich eingebaut werden.

In [62]:
def alt_jung(alter1,alter2):
    if alter1<alter2:
        return "jünger"
    else:
        return "älter"
name,alter="Jonas",28
name2,alter2="Karl",32
F"{name} ist {abs(alter-alter2)} Jahre {alt_jung(alter,alter2)} als {name2}."

'Jonas ist 4 Jahre jünger als Karl.'

Bisher haben wir zwar strings in der Ausgabe zusammengesetzt, aber die Ausgabe der einzelnen Elemente ist nicht ordentlich gestltet und dabei geht es ja bei der formatierten Ausgabe. Wir werden das jetzt für die f-strings einführen, aber vorher noch ein Beispiel, warum formatierte Ausgabe Sinn macht.

In [63]:
einkäufe=[
        (7,"Packungen Tomaten", 3.76),
        (12, "Nudeln", 1.45),
        (5,"Bio Rindfleisch spezial", 24.20),
        ]
kurs_euro_sfr=1.08
for anzahl, text, preis in einkäufe:
    gesamt= anzahl*preis
    gesamt_in_sfr=gesamt*kurs_euro_sfr
    print(f"{anzahl} von {text} in SFR {gesamt_in_sfr}")

7 von Packungen Tomaten in SFR 28.425600000000003
12 von Nudeln in SFR 18.791999999999998
5 von Bio Rindfleisch spezial in SFR 130.68


Das ist um es gelinde auszudrücken nicht sehr user-freundlich. Hier ergänzen wir unser f-strings mit einer Angabe, wie und wo die eingefügten Ausdrücke ausgegeben werden sollen. Ein Beispiel für unseren obigen Fall:

In [64]:
einkäufe=[
        (7,"Packungen Tomaten", 3.76),
        (12, "Nudeln", 1.45),
        (5,"Bio Rindfleisch spezial", 24.20),
        ]
kurs_euro_sfr=1.08
for anzahl, text, preis in einkäufe:
    gesamt= anzahl*preis
    gesamt_in_sfr=gesamt*kurs_euro_sfr
    print(f"{anzahl:4d} von {text:25s} in SFR {gesamt_in_sfr:6.2F}")

   7 von Packungen Tomaten         in SFR  28.43
  12 von Nudeln                    in SFR  18.79
   5 von Bio Rindfleisch spezial   in SFR 130.68


Sieht doch schon viel besser aus! Vor Schliessen der geschweiften Klammer eines Ausdrucks können wir nach einem Doppelpunkt auf diese Weise noch spezifizieren, wieviel Platz die Ausgabe haben soll, welcher Typ die Ausgabe ist (d oder D-> integer, f oder F-> Fliesskommazahl, s oder S-> String, E oder e-> Exponentialzahl, X oder x-> Hexadezimalzahl. Bei Fliesskommazahlen oder Exponentialzahlen können darüber hinaus die Nachkommastellen angegeben werden. <br>Wenn wir uns unsere Artikelbezeichnung anschauen, wäre es vielleicht schön, wenn diese zentriert ausgegeben würden. Dies lässt sich durch ein entsprechendes Zeichen für die Ausrichtung erreichen.

In [65]:
einkäufe=[
        (7,"Packungen Tomaten", 3.76),
        (12, "Nudeln", 1.45),
        (5,"Bio Rindfleisch spezial", 24.20),
        ]
kurs_euro_sfr=1.08
for anzahl, text, preis in einkäufe:
    gesamt= anzahl*preis
    gesamt_in_sfr=gesamt*kurs_euro_sfr
    print(f"{anzahl:4d} von {text:^25s} in SFR {gesamt_in_sfr:6.2f}")

   7 von     Packungen Tomaten     in SFR  28.43
  12 von          Nudeln           in SFR  18.79
   5 von  Bio Rindfleisch spezial  in SFR 130.68


Folgende Optionen sind hierfür einsetzbar:<br>
 

<table cellpadding="12" cellspacing="12" border="1" bgcolor="#F5F5F5">
  <thead>
    <tr>
      <th><center>Ausrichtungszeichen</th>
      <th><center>Bedeutung</th>
      </tr>
    </thead>
  <tbody>
    <tr><td valign="baseline"><tt><center><</tt></td>
        <td><left><center>linksbündige Ausgabe (für strings default)</td>
        </tr>
    <tr><td valign="baseline"><tt><center>></tt></td>
        <td><center>rechtsbündige Ausgabe (für numerische Werte default).</td>
        </tr>
    <tr><td valign="baseline"><tt><center>^     </tt></td>
        <td><center>zentrierte Ausgabe</td>
        </tr>
    <tr><td valign="baseline"><tt><center>=</tt></td>
        <td><center>der Platz zwischen dem Vorzeichen (falls vorhanden, sonst Beginn des Feldes)<br> und dem Beginn der Zahl mit Auffüllzeichen belegen (s.u.) </td>
        </tr>
    <tr><td valign="baseline"><tt><center>+</tt></td>
        <td><center>Vorzeichen wird immer ausgegeben</td>
        </tr>
    <tr><td valign="baseline"><tt><center>-</tt></td>
        <td><center>Vorzeichen nur bei negativem Ausdruck</td>
        </tr>
    <tr><td valign="baseline"><tt><center>Leerzeichen</tt></td>
        <td><center>Ausdruck positiv -> Leerzeichen, Ausdruck negativ -> - Zeichen</td>
        </tr>
        </tbody>
</table>

In [66]:
einkäufe=[
        (7,"Packungen Tomaten", 3.76),
        (12, "Nudeln", 1.45),
        (5,"Bio Rindfleisch spezial", 2477.20),
        ]
kurs_euro_sfr=1.08
for anzahl, text, preis in einkäufe:
    gesamt= anzahl*preis
    gesamt_in_sfr=gesamt*kurs_euro_sfr
    print(f"{anzahl:004d} von {text:^25s} in SFR {gesamt_in_sfr:=006.2f}")

0007 von     Packungen Tomaten     in SFR 028.43
0012 von          Nudeln           in SFR 018.79
0005 von  Bio Rindfleisch spezial  in SFR 13376.88


Falls die Gesamtbreite zur Darstellung nicht ausreicht , wird wie oben sichtbar unsere Formatierung logischerweise zerstört.

### Die alte Methode im Stil von printf und sprintf

### Statt "alte" Methode sollten wir besser "veraltete" Methode sagen, da dieses Vorgehen keinen guten Python-Stil mehr darstellt.

Gibt es ein printf in Python? Das ist die brennende Frage für Python-Anfänger, die von C, Perl, Bash oder anderen Programmiersprachen kommen, die dieses Statement bzw. diese Funktion benutzen. Zu sagen, dass Python eine print-Funktion und keine printf-Funktion hat ist nur die eine Seite der Medaille. Also gibt es doch eine "printf"-Funktion in Python? Nein, aber die Funktionalität des "alten" printf wurde in anderer Form übernommen. Zu diesem Zweck wird der Modulo-Operator "%" von der String-Klasse überladen. Deshalb spricht man bei dem überladenen Operator "%" der String-Klasse auch häufig vom String-Modulo-Operator, obwohl es eigentlich wenig mit der eigentlich Modulo-Arithmetik zu tun hat. Manchmal wird er auch synonym als String-Modulus-Operator genannt. Ein weiterer Begriff, für diese Operation lautet "String-Interpolation", weil bei dieser Operation Werte, wie Integers, Floats usw., in den zu erzeugenden formatierten String eingefügt also interpoliert werden. Den mittels der String-Interpolation erzeugten formatierten String, kann man dann z.B. mittels print (also nicht printf) ausgeben. Aber man kann diesen formatierten String auch auf andere Art im Programm weiter verwenden, wie ihn zum Beispiel in eine Datenbank übertragen.
Seit Python 2.6 eingeführt worden ist, wird jedoch von der Python-Gemeinde empfohlen die String-Methode "format" stattdessen zu verwenden. Unglücklicherweise konnten man sich auch bei dem Design von Python3 nicht von der veralteten String-Interpolations-Methode trennen. Außerdem wird die veraltete Methode noch allzu häufig benutzt. Aus diesem Grunde gehen wir auch in unserem Tutorial ausgiebig darauf ein. Um fremden Python-Code verstehen zu können, der die String-Interpolations-Methode benutzt, ist es nötig den Mechanismus zu verstehen. Wahrscheinlich - oder hoffentlich - wird man sich jedoch bei der Weiterentwicklung von Python eines Tages von dieser Methode trennen.

Das folgende Diagramm verbildlicht die Arbeitsweise des String-Modulo-Operators:

<img src="../images/formatierte_ausgabe_beispiel.webp" width="500" alt="Formatierte Ausgabe mittels String-Modulo-Operator an einem Beispiel" />

Auf der linken Seite des String-Modulo-Operators befindet sich der sogenannte Format-String und auf der rechten Seite steht ein Tupel mit den Werten, die in den Format-String interpoliert bzw. eingefügt werden sollen. Die Werte können Literale, Variablen oder beliebige arithmetische Ausdrücke sein.

<img src="../images/formatstring.webp" width="500" alt="Formatstring und Wertetupel" />

Der Formatstring enthält Platzhalter. In unserem Beispiel befinden sich zwei davon: "%5d" und "%8.2f".

Die allgemeine Syntax für diese Platzhalter lautet:

<pre>
    %[flags][width][.precision]type 
</pre>


<img  class="imgright" src="../images/string_modulo_float.webp" srcset="../images/string_modulo_float_350w.webp 350w,../images/string_modulo_float_300w.webp 300w" width="200" alt="Float-Format" />

Schauen wir uns einmal die Platzhalter unseres Beispieles genauer an: Der zweite, also "%8.2f", ist eine Formatbeschreibung für eine Floatzahl. Wie auch alle anderen Platzhalter wird er durch ein "%"-Zeichen eingeleitet. Dieses wird von einer Zahl gefolgt, die die totale Anzahl der Zeichen angibt, mit der die Zahl ausgegeben werden soll. Diese Anzahl beinhaltet auch den Dezimalpunkt und alle Ziffern vor und nach dem Punkt. Unsere Floatzahl 59.058 muss also mit 8 Zeichen dargestellt werden. Der Dezimalanteil der Zahl ist auf 2 gesetzt, dies ist die Zahl hinter dem "." beim Platzhalter. Das bedeutet, dass die Anzahl der Nachkommastellen auf 2 gesetzt ist. Das letzte Zeichen im Platzhalt ist der Buchstabe "f". Wie sich leicht vermuten lässt, steht er für "float".

Schaut man sich die Ausgabe an, erkennt man, dass die Zahl "59.058" als " 59.06" ausgegeben wird. Man sieht auch, dass die Ausgabe nicht etwa nach zwei Stellen abgebrochen wird, sondern dass eine Rundung erfolgt. Außerdem werden von links drei Leerzeichen an die Zahl bei der Ausgabe angefügt.

Der erste Platzhalter "%5d" dient als Beschreibung für die erste Komponente unseres Tupels, d.h. die Ganzzahl (integer) 453. Diese Zahl soll mit 5 Zeichen ausgedruckt werden. Da 453 aber nur aus 3 Stellen besteht, werden der Zahl in der Ausgabe zwei Leerzeichen vorangestellt. Viele werden statt "%5d" einen Platzhalter der Art "%5i" erwartet haben, da es sich ja um eine Integer-Zahl handelt. Worin besteht der Unterschied? Ganz einfach: Es gibt keinen Unterschied zwischen "d" und "i". Beide werden zur Formatierung von Integers genutzt.

Die Vorteile bzw. die Schönheit der formatierten Ausgabe kann man nur ermessen, wenn man mehrere gleichermaßen formatierte Ausgaben ausgibt. Im folgenden Beispiel zeigen wir, wie es aussieht, wenn wir 5 Fließzahlen (floats) gleich formatiert mit dem Platzhalter "%6.2f" in verschiedenen Zeilen ausgeben:


<img   src="../images/formatting_floats.webp" srcset="../images/formatting_floats_800w.webp 800w,../images/formatting_floats_700w.webp 700w,../images/formatting_floats_600w.webp 600w,../images/formatting_floats_500w.webp 500w,../images/formatting_floats_400w.webp 400w,../images/formatting_floats_350w.webp 350w,../images/formatting_floats_300w.webp 300w"  width="450" alt="Formatierte Floatzahlen in verschiedenen Zeilen" />

<table cellpadding="6" cellspacing="0" border="1" bgcolor="#F5F5F5">
  <thead>
    <tr>
      <th>Platzhaltersymbol</th>
      <th>Bedeutung</th>
      </tr>
    </thead>
  <tbody>
    <tr><td valign="baseline"><tt>d</tt></td>
        <td>Vorzeichenbehaftete Ganzzahl (Integer, dezimal).</td>
        </tr>
    <tr><td valign="baseline"><tt>i</tt></td>
        <td>Vorzeichenbehaftete Ganzzahl (Integer, dezimal).</td>
        </tr>
    <tr><td valign="baseline"><tt>o</tt></td>
        <td>vorzeichenlose Ganzzahlen (oktal)</td>
        </tr>
    <tr><td valign="baseline"><tt>u</tt></td>
        <td>vorzeichenlose Ganzzahlen (dezimal)</td>
        </tr>
    <tr><td valign="baseline"><tt>x</tt></td>
        <td>vorzeichenlose Ganzzahlen (hexadezimal)</td>
        </tr>
    <tr><td valign="baseline"><tt>X</tt></td>
        <td>vorzeichenlose Ganzzahlen (hexadezimal), Uppercase</td>
        </tr>
    <tr><td valign="baseline"><tt>e</tt></td>
        <td>Fließkommazahl im Exponentialformat (in Kleinbuchstaben).</td>
        </tr>
    <tr><td valign="baseline"><tt>E</tt></td>
        <td>Fließkommazahl im Exponentialformat (in Großbuchstaben).</td>
        </tr>
    <tr><td valign="baseline"><tt>f</tt></td>
        <td>wie e</td>
        </tr>
    <tr><td valign="baseline"><tt>F</tt></td>
        <td>wie E</td>
        </tr>
    <tr><td valign="baseline"><tt>g</tt></td>
        <td>Wie "<tt>e</tt>", wenn der Exponent größer ist als -4 oder
              oder kleiner als die Präzision. Ansonsten wie "<tt>f</tt>"</td>
        </tr>
    <tr><td valign="baseline"><tt>G</tt></td>
        <td>wie "<tt>E</tt>" und analog zu "<tt>g</tt>"</td>
        </tr>
    <tr><td valign="baseline"><tt>c</tt></td>
        <td>ein Zeichen</td>
        </tr>
    <tr><td valign="baseline"><tt>s</tt></td>
        <td>Eine Zeichenkette (String); beliebige Python-Objekte werden in String mittels der Methode <tt>str()</tt>) gewandelt.</td>
        </tr>
    <tr><td valign="baseline"><tt>r</tt></td>
        <td>siehe <tt>s</tt></td>
        </tr>
    <tr><td valign="baseline"><tt>%</tt></td>
        <td>Es findet keine Argument-Konvertierung statt, es wird ein "<tt>%</tt>"-Zeichen ausgegeben.</td>
        </tbody>
</table>

Die folgenden Beispiele zeigen einige exemplarische Anwendungen der Konvertierungsmöglichkeiten der vorigen Tabelle:

In [67]:
print("%10.3e"% (356.08977))

 3.561e+02


In [68]:
print("%10.3E"% (356.08977))

 3.561E+02


In [69]:
print("%10o"% (25))

        31


In [70]:
print("%10.3o"% (25))

       031


In [71]:
print("%10.5o"% (25))

     00031


In [72]:
print("%5x"% (47))

   2f


In [73]:
print("%5.4x"% (47))

 002f


In [74]:
print("%5.4X"% (47))

 002F


In [75]:
print("Ein Prozentzeichen: %% " % ())

Ein Prozentzeichen: % 


<table cellpadding="6" cellspacing="0" border="1" bgcolor="#F5F5F5">
  <thead>
    <tr>
      <th>Flag</th>
      <th>Bedeutung</th>
      </tr>
    </thead>
  <tbody>
    <tr><td valign="baseline"><tt>#</tt></td>
        <td>Wird dieses Zeichen mit o, x oder X benutzt, wird der jeweilige Wert mit dem entsprechenden folgenden Präfix: 0, 0o, 0O, 0x oder 0X
</td></tr>
    <tr><td valign="baseline"><tt>0</tt></td>
        <td>Das Ergebnis der Umwandlung wird mit Nullen aufgefüllt.</td></tr>
    <tr><td valign="baseline"><tt>-</tt></td>
        <td>Das Ergebnis der Umwandlung wird linksbündig ausgegeben.</td></tr>
    <tr><td valign="baseline"><tt>&nbsp;</tt></td>
        <td>Falls kein Vorzeichen (beispielsweise ein Minuszeichen) ausgegeben wird, wird ein Leerzeichen vor den Wert gesetzt.</td></tr>
    <tr><td valign="baseline"><tt>+</tt></td>
        <td>Das Ergebnis der Umwandlung wird mit einem Vorzeichen versehen ("<tt>+</tt>" oder "<tt>-</tt>"). Dieses Flag überschreibt ein "space"-Flag).</td></tr></tbody>
</table>

Beispiele:

In [76]:
print("%#5X"% (47))

 0X2F


In [77]:
print("%5X"% (47))

   2F


In [78]:
print("%#5.4X"% (47))

0X002F


In [79]:
print("%#5o"% (25))

 0o31


In [80]:
print("%+d"% (42))

+42


In [81]:
print("% d"% (42))

 42


In [82]:
print("%+2d"% (42))

+42


In [83]:
print("% 2d"% (42))

 42


In [84]:
print("%2d"% (42))

42


Obwohl es so ausschauen mag, ist die Formatierung nicht Teil der Print-Funktion. Schaut man sich die vorigen Beispiele genauer an, sieht man, dass wir einen formatierten String an die Print-Funktion übergeben haben. Oder anders ausgedrückt: Wenn die String-Interpolation auf einen String angewendet wird, liefert sie einen String zurück. Dieser String wird dann an print übergeben. Dies bedeutet wiederum, dass wir die String-Interpolation auch hätten "zweistufig" anwenden können, d.h. wir hätten erst einen formatierten String erzeugt, diesen einer Variablen zugewiesen und diese Variable dann an print übergeben:

In [85]:
s = "Preis: $ %8.2f"% (356.08977)
print(s)

Preis: $   356.09


### Der alte pythonische Weg: Die String-Methode "format"

Was die String-Methode format betrifft, ist die help-Funktion von Python nicht sehr hilfreich. Alles was sie uns sagt, ist dieses:

 |  format(...)
 <br>
 |      S.format(*args, **kwargs) -> str
 <br>
 |      
 <br>
 |      Return a formatted version of S, using substitutions from args and kwargs.
 <br>
 |      The substitutions are identified by braces ('{' and '}').
 <br>
 |  

Die String-Methode "format" wurde mit Python 2.6 als Ersatz für die String-Interpolation eingeführt.

Ganz allgemein sieht "format" wie folgt aus:

<pre>
template.format(p0, p1, ..., k0=v0, k1=v1, ...)
</pre>

Der Format-String ist ein String, der ein oder mehrere Format-Codes innerhalb eines konstanten Textes enthält, d.h. Felder, die ersetzt werden sollen. Die zu "ersetzenden Felder" sind von geschweiften Klammern umgeben. Eine geschweifte Klammer und der von ihr umschlossene "Code" wird mit dem formatierten Wert aus dem korrespondierenden Argument ersetzt. Nach welchen Regeln dies zu erfolgen hat, werden wir im folgenden erläutern. Alles andere, was nicht von geschweiften Klammern umschlossen ist, wird wörtlich, also ohne Änderungen, ausgedruckt. Möchte man eine öffnende oder eine schließende geschweifte Klammer ausgeben, muss man diese verdoppeln, also "{{" oder "}}".

Es gibt zwei Arten von Argumenten für die format-Methode. Die Argumentenliste startet mit 0 oder mehr Positionsargumenten (p0, p1, ...), die von 0 oder mehr Schlüsselwortargumenten der Form name=value gefolgt werden können.

Ein Positionsparameter der format-Methode kann benutzt werden, indem man den Index des Parameters nach öffnenden geschweiften Klammer angibt, d.h. {0} für den ersten Parameter, {1} für den zweiten und so weiter. Der Index des Positionsparameter nach der öffnenden geschweiften Klammer kann von einem Doppelpunkt und einem Formatstring gefolgt werden, der dem des String-Interpolationsparameters ähnelt, den wir zu Beginn des Kapitels besprochen hatten, also zum Beispiel {0:5d}
Falls die Positionsparameter in der Reihenfolge benutzt werden, in der sie in der Parameterliste stehen, kann die Angabe des Indexes innerhalb der geschweiften Klammern entfallen, d.h. '{} {} {}' entspricht '{0} {1} {2}'. Um es nochmals klar zu stellen: Will man sie in einer anderen Reihenfolge benutzen, sind die Angaben der Indexpositionen dringend notwendig, wie z.B. hier '{2} {1} {0}'

Im folgenden Diagramm zeigen wir anhand eines Beispieles wie die Stringmethode "format" für zwei Parameter funktioniert:

<img src="../images/format_method_positional_parameters.webp" height="190" alt="Allgemeine Arbeitsweise der Stringmethode format für zwei Positionsparameter" />

Beispiele mit Positionsparametern:

In [86]:
"Erstes Argument: {0}, zweites: {1}".format(47,11) 

'Erstes Argument: 47, zweites: 11'

In [87]:
"Zweites Argument: {1}, erstes: {0}".format(47,11) 

'Zweites Argument: 11, erstes: 47'

In [88]:
"Zweites Argument: {1:3d}, erstes: {0:7.2f}".format(47.42,11) 

'Zweites Argument:  11, erstes:   47.42'

In [89]:
"Erstes Argument: {}, zweites: {}".format(47,11) 

# Argumente können auch mehrmals verwendet werden:

'Erstes Argument: 47, zweites: 11'

In [90]:
"precisions: {0:6.2f} or {0:6.3f}".format(1.4148) 

'precisions:   1.41 or  1.415'

Im folgenden Beispiel demonstrieren wir, wie Schlüsselwortparameter mit der format-Methode benutzt werden können:

In [91]:
"Artikel: {a:5d},  Preis: {p:8.2f}".format(a=453, p=59.058)

'Artikel:   453,  Preis:    59.06'

<img src="../images/format_method_keyword_parameters.webp" height="190" alt="Allgemeine Arbeitsweise der Stringmethode format für zwei Schlüsselwortparameter" />


Mit der format-Methode ist es auch möglich Daten links- und rechtsbündig auszugeben. Dazu müssen wir der Formatierungsanweisung ein "<" (linksbündig) oder ein ">" (rechtsbündig) voranstellen. Wir demonstrieren dies an den folgenden Beispielausgaben:

In [92]:
"{0:<20s} {1:6.2f}".format('Spam & Eggs:', 6.99)

'Spam & Eggs:           6.99'

In [93]:
"{0:>20s} {1:6.2f}".format('Spam & Ham:', 7.99)

'         Spam & Ham:   7.99'

In [94]:
"{0:<20} {1:6.2f}".format('Spam ', 0)

'Spam                   0.00'

If the width field is preceded by a zero ('0') character,  
sign-aware zero-padding for numeric types will be enabled.

In [95]:
x = 378
print("The value is {0:06d}".format(x))

The value is 000378


In [96]:
x = -378
print("The value is {0:06d}".format(x))

The value is -00378


In [97]:
"{0:>20} {1:6.2f}".format('Spam & Ham:', 7.99)

'         Spam & Ham:   7.99'

<table  cellpadding="6" cellspacing="0" border="1" bgcolor="#F5F5F5">
<colgroup>
<col width="14%" />
<col width="88%" />
</colgroup>
<thead valign="bottom">
<tr><th class="head">Ausrichtungs-Option</th>
<th class="head">Bedeutung</th>
</tr>
</thead>
<tbody valign="top">
<tr><td><tt><span class="pre">'&lt;'</span></tt></td>
<td>Das Feld wird linksbündig innerhalb des vorhandenen Raumes ausgegeben. Strings werden standardmäßig linksbündig ausgegeben.</td>
</tr>
<tr><td><tt><span class="pre">'&gt;'</span></tt></td>
<td>Das Feld wird rechtsbündig innerhalb des vorhandenen Raumes ausgegeben. Numerische Werte werden standardmäßig rechtsbündig ausgegeben.</td>
</tr>

<tr><td><tt><span class="pre">'0'</span></tt></td>
<td>Wenn man das Formatfeld mit einer führenden Null ('0') versieht, erfolgt ein 
Auffüllen mit Nullen unter Beachtung eines möglichen Vorzeichens.
<pre>
x = 378
print("The value is {0:06d}".format(x))
The value is 000378
x = -378
print("The value is {0:06d}".format(x))
The value is -00378
</pre>
</td>
</tr>
<tr><td><tt><span class="pre">','</span></tt></td>
<td>Mit dieser Option erhält man Tausender Gruppierungen, die mit Komma getrennt sind: 
<pre>
print("The value is {:,}".format(x))
The value is 78,962,324,245
print("The value is {0:6,d}".format(x))
The value is 5,897,653,423
x = 5897653423.89676
print("The value is {0:12,.3f}".format(x))
The value is 5,897,653,423.897
</pre>
</td>
</tr>


<tr><td><tt><span class="pre">'='</span></tt></td>
<td>Die Auffüllzeichen (padding characters) werden zwischen Vorzeichen, falls ein Vorzeichen ausgegeben wird und dem eigentlichen Beginn der Ziffern einer Zahl eingeführt. Damit können Felder der Art "+000000120" ausgegeben werden. Diese Ausrichtungsoption kann nur auf numerische Werte angewendet werden.</td>
</tr>
<tr><td><tt><span class="pre">'^'</span></tt></td>
<td>Ein Feld wird zentriert innerhalb des zur Verfügung stehenden Raumes ausgegeben.</td>
</tr>
</tbody>
</table>

Wenn keine minimale Feldlänge angegeben wird, entspricht die Feldlänge immer der Länge der Daten, die ein Feld enthält. In diesen Fällen haben die Ausrichtungsoptionen keine Bedeutung.

Zusätzlich können wir noch die Ausgabe mittels der sign-Option beeinflussen. Diese Optionen sind nur für numerische Werte gültig:

<table cellpadding="6" cellspacing="0" border="1" bgcolor="#F5F5F5">
<colgroup>
<col width="14%" />
<col width="88%" />
</colgroup>
<thead valign="bottom">
<tr><th class="head">sign-Option</th>
<th class="head">Bedeutung</th>
</tr>
</thead>
<tbody valign="top">
<tr><td><tt><span class="pre">'+'</span></tt></td>
<td>Es soll immer ein Vorzeichen ausgegeben werden, also unabhängig davon, ob es sich um eine positive oder negative Zahl handelt.</td>
</tr>
<tr><td><tt><span class="pre">'-'</span></tt></td>
<td>Ein Vorzeichen soll nur bei negativen Zahlen verwendet werden.</td>
</tr>
<tr><td>space</td>
<td>Statt eines "+"-Zeichens wird bei positiven Zahlen ein Leerzeichen " " vorangestellt. Bei negativen Zahlen ein Minus-Zeichen "-".</td>
</tr>
</tbody>
</table>

### Benutzung von Dictionaries beim Aufruf der "format"-Methode
In den vorigen Kapitel haben wir gesehen, dass wir zwei Möglichkeiten haben Werte mit "format" zu formatieren:

- Benutzung des Index bzw. der Position des Argumentes:


In [98]:
print("Die Hauptstadt von {0:s} ist {1:s}".format("Ontario","Toronto"))

Die Hauptstadt von Ontario ist Toronto


Um es nochmals zu erwähnen: Im vorigen Beispiel hätten wir auch geschweifte Klammern mit leeren Inhalt benutzen können!
- Benutzung von Schlüsselwort-Parametern:

In [99]:
print("Die Hauptstadt von {province} ist {capital}".format(province="Ontario",capital="Toronto"))    

Die Hauptstadt von Ontario ist Toronto


Im zweiten Fall könnten wir auch ein Dictionary verwenden, wie wir im folgenden Code sehen können:

In [100]:
data = dict(province="Ontario",capital="Toronto")
data

{'province': 'Ontario', 'capital': 'Toronto'}

In [101]:
print("The capital of {province} is {capital}".format(**data))

The capital of Ontario is Toronto


Das doppelte Sternchen ("*") vor data wandelt data automatisch in die Form 'province="Ontario",capital="Toronto"'. Schauen wir uns das folgende Python-Programm an:

In [102]:
capital_country = {"Vereinigte Staaten" : "Washington", 
                   "US" : "Washington", 
                   "Kanada" : "Ottawa",
                   "Deutschland": "Berlin",
                   "Frankreich" : "Paris",
                   "England" : "London",
                   "Großbritannien" : "London",
                   "Schweiz" : "Bern",
                   "Österreich" : "Wien",
                   "Niederlande" : "Amsterdam"}

print("Länder und Ihre Hauptstädte:")
for c in capital_country:
    print("{country}: {capital}".format(country=c, capital=capital_country[c]))    

Länder und Ihre Hauptstädte:
Vereinigte Staaten: Washington
US: Washington
Kanada: Ottawa
Deutschland: Berlin
Frankreich: Paris
England: London
Großbritannien: London
Schweiz: Bern
Österreich: Wien
Niederlande: Amsterdam


Das vorige Programm können wir so umschreiben, dass im Aufruf der format-Methode das Dictionary direkt verwendet wird.

In [103]:
capital_country = {"Vereinigte Staaten" : "Washington", 
                   "US" : "Washington", 
                   "Kanada" : "Ottawa",
                   "Deutschland": "Berlin",
                   "Frankreich" : "Paris",
                   "England" : "London",
                   "Großbritannien" : "London",
                   "Schweiz" : "Bern",
                   "Österreich" : "Wien",
                   "Niederlande" : "Amsterdam"}

print("Countries and their capitals:")
for c in capital_country:
    format_string = c + ": {" + c + "}" 
    print(format_string.format(**capital_country))

Countries and their capitals:
Vereinigte Staaten: Washington
US: Washington
Kanada: Ottawa
Deutschland: Berlin
Frankreich: Paris
England: London
Großbritannien: London
Schweiz: Bern
Österreich: Wien
Niederlande: Amsterdam


Der Ausdruck ist gleich wie beim ersten Programm.

### Benutzung von lokalen Variablen in "format"
"locals" ist eine Funktion, die ein Dictionary mit den lokal definierten Namen und ihren aktuellen Werten zurückliefert. Die Variablen sind die Schlüssel dieses Dictionary und die Werte zu den Schlüsseln entsprechen den Werten der Variablen:

In [104]:
a = 42
b = 47
def f(): return 42
locals()

{'__name__': '__main__',
 '__doc__': 'Automatically created module for IPython interactive environment',
 '__package__': None,
 '__loader__': None,
 '__spec__': None,
 '__builtin__': <module 'builtins' (built-in)>,
 '__builtins__': <module 'builtins' (built-in)>,
 '_ih': ['',
  'Formatierte Stringliterale oder f-strings wurden ab Python 3.6 eingeführt. Syntax ist sehr einfach und flexibel. Ein Beispiel ist:',
  'name,alter ="Jonas",19\nf"Mein Freund {Jonas} ist {alter} Jahre alt"',
  'name,alter ="Jonas",19\nf"Mein Freund {name} ist {alter} Jahre alt"',
  'name,alter ="Jonas",19\nf"Mein Freund {name} ist {alter} Jahre alt"\nprint(f"Mein Freund {name} ist {alter} Jahre alt")',
  'name,alter ="Jonas",19\nf"Mein Freund {name} ist {alter} Jahre alt"',
  'name,alter ="Jonas",19\nf"Mein Freund {name} ist {alter} Jahre alt"\nprint(f"Mein Freund {name} ist {alter} Jahre alt")',
  'name,alter="Jonas",23\nname2,alter2="Kartl",32\nf"{name} ist {abs(alter-alter2)} Jahre {"jünger" if alter-alter2<0

Das Dictionary, was von locals() zurückgeliegert wird, kann als Parameter für die format-Methode genutzt werden. Dadurch ist es möglich im Format-String lokale Variablen zu verwenden.

Im folgenden Beispiel benutzen wir die lokalen Variablen a, b und f:

In [105]:
print("a={a}, b={b} and f={f}".format(**locals()))

a=42, b=47 and f=<function f at 0x000000FD759E39D0>


### Weitere String-Methoden zum Formatieren
Die String-Klasse enthält weitere Methoden, die man auch zu Formatierungszwecken nutzen kann: ljust, rjust, center und zfill.

Falls S ein String ist, können wir die vier Methoden wie folgt erklären:

- center(...):
<pre>
    S.center(width[, fillchar]) -> str
</pre>
An den String S werden links und rechts Füllzeichen bis zur Länge "width" so angefügt, dass der ursprüngliche String zentriert wird. Wird für das Füllzeichen "fillchar" kein Wert angegeben, wird als Standardwert ein Leerzeichen verwendet.

*Beispiele: 

In [106]:
s = "Python"
s.center(10)   

'  Python  '

In [107]:
s.center(10,"*") 

'**Python**'

- ljust(...):
<pre>
     S.ljust(width[, fillchar]) -> str 
</pre>
Der String S wird linksbündig zurückgeliefert. Rechts wird der String mit Füllzeichen "fillchar" bis zur Länge "width" angefüllt. Auch hier ist ein Leerzeichen der Default-Wert für "fillchar".

*Beispiele:

In [108]:
s = "Training"
s.ljust(12)

'Training    '

In [109]:
s.ljust(12,":")

'Training::::'

- rjust(...):

<pre>
    S.rjust(width[, fillchar]) -> str
</pre>
    
Der String S wird rechtsbündig zurückgeliefert. Von links wird der String mit Füllzeichen "fillchar" bis zur Länge "width" angefüllt. Auch hier ist ein Leerzeichen der Default-Wert für "fillchar".

*Beispiele:

In [110]:
s = "Programming"
s.rjust(15)

'    Programming'

In [111]:
s.rjust(15, "~")

'~~~~Programming'

- zfill(...):

<pre>    
    S.zfill(width) -> str
</pre>

Ein String S wird von links mit Nullen bis zu der Länge "width" aufgefüllt. Der String " wird nicht abgeschnitten, falls er bereits länger als die vorgesehen Länge "width" ist. Diese Methode entspricht einem S.rjust(width, "0")

*Beispiele:

In [112]:
account_number = "43447879"
account_number.zfill(12)

'000043447879'

In [113]:
# can be emulated with rjust:

account_number.rjust(12,"0")

'000043447879'